### ISSUE TITLE
Premultiplied vs non-premultiplied alpha in HEIF

### ISSUE THREAD

**Issue by jonsneyers** [strukturag/libheif#471]
In HEIF/HEIC, there is an optional way to signal alpha channel information.

In particular, there's a field called `alpha_channel_use_idc`.

From the HEIF spec:
_"alpha_channel_use_idc equal to 0 indicates that for alpha blending purposes the decoded samples of the associated primary picture should be multiplied by the interpretation sample values of the auxiliary coded picture in the display process after output from the decoding process. alpha_channel_use_idc equal to 1 indicates that for alpha blending purposes the decoded samples of the associated primary picture should not be multiplied by the interpretation sample values of the auxiliary coded picture in the display process after output from the decoding process. alpha_channel_use_idc equal to 2 indicates that the usage of the auxiliary picture is unspecified. Values greater than 2 for alpha_channel_use_idc are reserved for future use by ITU-T | ISO/IEC. When not present, the value of alpha_channel_use_idc is inferred to be equal to 2."_

As far as I can tell, libheif is currently not signaling any alpha channel information (it is optional). This means that it is up to applications to decide whether they want to interpret the alpha in a premultiplied (associated) way or in a non-premultiplied way (unassociated, like in PNG). That is obviously not a good situation: it does look like at least in some cases, iOS is interpreting things in the premultiplied way, while other applications are interpreting things in the PNG way. Needless to say that it is very undesirable to have an image with alpha which can be interpreted/blended in two very different ways.

So I would suggest to _always_ signal this field (to be explicitly set to non-premultiplied, i.e.`alpha_channel_use_idc == 1`), since that is the only way to ensure consistency of decoded images.

**Comment by farindk** [comment]
Yes, that makes sense. But where is this defined? I only have ISO 23008-12, 2017-12 and cannot find anything about this in that document.

**Comment by farindk** [comment]
Ok, I have found it. It's not in the HEIF standard but in a SEI in h.265.
Thus, we'll need support for that SEI in libde265 first.

**Comment by jonsneyers** [comment]
Yes. It's OK to not implement premultiplied encoding/decoding for now, but it is important at this point to signal "not premultiplied" when encoding and maybe throw an error when decoding something that is explicitly signaled to be premultiplied.

(AVIF has a similar thing btw but they do it via an `iref` of type `prem`)

**Comment by Jamaika1** [comment]
I have a question. Why is it impossible to convert the alpha channel from PNG to AVIF in version AV1 3.0.0. The specification has changed. 
heifenc.exe image_21447_32bit.png -v -L -q 100 --avif -p chroma=444 -p speed=5 -p tune=ssim --matrix_coefficients=0 --full_range_flag=1 --enable-two-colr-boxes -o image_21447_32bit.avif

**Comment by farindk** [comment]
The `iref` reference of type `prem` is part of the MIAF standard, so it should apply to both HEIC and AVIF.
As far as I understand the MIAF standard, there is the usual `auxl` link to the master image. By default, the semantics is non-premultiplied. If there is another `prem` link in the other direction (master to alpha image), the semantics change to pre-multiplied.

**Comment by baumanj** [comment]
> The `iref` reference of type `prem` is part of the MIAF standard, so it should apply to both HEIC and AVIF.
> As far as I understand the MIAF standard, there is the usual `auxl` link to the master image. By default, the semantics is non-premultiplied. If there is another `prem` link in the other direction (master to alpha image), the semantics change to pre-multiplied.

This agrees with what we've implemented in [mp4parse-rust](https://github.com/mozilla/mp4parse-rust/blob/b4cc8c4f3fd5bf1cf9a5ecc5768d66e6dc630051/mp4parse/src/lib.rs#L1513-L1519) (the parser used by Firefox). Hat tip to @kornelski for implementing it.

**Comment by jonsneyers** [comment]
Having a way to unambiguously signal non-premultiplied alpha would also be nice (and doing this by default would make sense too).

**Comment by kornelski** [comment]
Why would you signal that? It's just a bloat in the file. It'd imply that lack of signal doesn't mean "definitely not premultiplied, no doubt about it" (which is true), but rather suggest there's an ambiguity.

**Comment by farindk** [comment]
Yes, MIAF says that alpha is premultiplied when there is a `prem` reference and not-premultiplied otherwise. So I do not see why another flag should be introduced that can lead to inconsistencies.

Still, the HEIF standard recommends that the alpha-plane h265 stream should contains an alpha description SEI message. As it only "recommends" it, I consider this optional and libheif does not send this currently. It can be discussed whether sending a SEI message makes sense. In my view, since this is optional, it cannot hold any mandatory information anyhow and will thus only lead to inconsistencies in how to interpret the data.
